appserver/security/realm-stores/src/main/java/fish/payara/security/realm/identitystores/PamRealmIdentityStore.java

```diff
@@ -0,0 +1,129 @@
+ *  elects to include this software in this distribution under the [CDDL or GPL
+ *  Version 2] license."  If you don't indicate a single choice of license, a
+ *  recipient has the option to distribute your version of this file under
+ *  either the CDDL, the GPL Version 2 or to extend the choice of license to
+ *  its licensees as provided above.  However, if you add GPL Version 2 code
+ *  and therefore, elected the GPL Version 2 license, then the option applies
+ *  only if the new code is made subject to such option by the copyright
+ *  holder.
+ */
+package fish.payara.security.realm.identitystores;
+
+import com.sun.enterprise.security.auth.login.common.LoginException;
+import com.sun.enterprise.security.auth.realm.NoSuchRealmException;
+import com.sun.enterprise.security.auth.realm.Realm;
+import com.sun.enterprise.security.auth.realm.pam.PamRealm;
+import com.sun.enterprise.security.ee.auth.login.PamLoginModule;
+import fish.payara.security.annotations.PamIdentityStoreDefinition;
+import fish.payara.security.realm.PamRealmIdentityStoreConfiguration;
+import fish.payara.security.realm.RealmUtil;
+import static fish.payara.security.realm.RealmUtil.ASSIGN_GROUPS;
+import static fish.payara.security.realm.RealmUtil.JAAS_CONTEXT;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Properties;
+import javax.enterprise.inject.Typed;
+import javax.security.enterprise.CallerPrincipal;
+import javax.security.enterprise.credential.Credential;
+import javax.security.enterprise.credential.UsernamePasswordCredential;
+import javax.security.enterprise.identitystore.CredentialValidationResult;
+import static javax.security.enterprise.identitystore.CredentialValidationResult.INVALID_RESULT;
+import static javax.security.enterprise.identitystore.CredentialValidationResult.NOT_VALIDATED_RESULT;
+import javax.security.enterprise.identitystore.IdentityStore;
+import org.glassfish.soteria.identitystores.IdentityStoreException;
+import org.jvnet.libpam.UnixUser;
+
+/**
+ * {@link PamRealmIdentityStore} Identity store validates the user using
+ * dynamically created pam realm instance and returns the validation result with
+ * the caller name and groups.
+ *
+ * @author Gaurav Gupta
+ */
+@Typed(PamRealmIdentityStore.class)
+public class PamRealmIdentityStore implements IdentityStore {
+
+    private PamRealmIdentityStoreConfiguration configuration;
+
+    private PamRealm pamRealm;
+
+    public static final Class<PamRealm> REALM_CLASS = PamRealm.class;
+
+    public static final Class<PamLoginModule> REALM_LOGIN_MODULE_CLASS = PamLoginModule.class;
+
+    public void init(PamIdentityStoreDefinition definition) {
+        configuration = PamRealmIdentityStoreConfiguration.from(definition);
+
+        try {
+            if (!Realm.isValidRealm(configuration.getName())) {
+                Properties props = new Properties();
+                props.put(JAAS_CONTEXT, configuration.getName());
```

Comment:
Same argument - default should be... oh there isn't predefined Login Module for pam, is it?

Reply:
Oh, there is - `pamRealm`